### ISSUE TITLE
unresolved-import with overlapping editable-installs

### ISSUE THREAD

**Issue by twmr** [astral-sh/ty#4371]
### Summary

When using editable installs with pip (using setuptools as build backend), ty fails to correctly resolve relative imports. This worked in ty 0.0.20, but fails since 0.0.21

```
error[unresolved-import]: Cannot resolve imported module `.`
 --> jobqueue/src/jobqueue/workflow.py:1:1
  |
1 | from . import utils
  | ^^^^^^^^^^^^^^^^^^^
```

I've finally managed to create a [MWE](https://gist.github.com/twmr/8e23285c6ba6dbc9fcd21547bb9d4d48), which you can easily run with different ty versions:

```
  TY_VERSION=0.0.20 ~/repro-ty-relative-import.sh # creates temp dir in /tmp
  TY_VERSION=0.0.21 ~/repro-ty-relative-import.sh
```

It could be that this is related to #475, but #475 was created before 0.0.21 was released in March.

Here are the last lines of the `ty check -vvv` output with ty 0.0.21

```
 2026-08-24T04:12:00.383515Z DEBUG ty_module_resolver::resolve: Adding editable installation to module resolution path /tmp/ty-relative-import-repro.qy3Ut3/jobqueue/src
    at crates/ty_module_resolver/src/resolve.rs:863 on ThreadId(2)
    in ty_project::check_file with file=file(Id(c00))
    in ty_project::Project::check

  2026-08-24T04:12:00.383801Z DEBUG ty_module_resolver::resolve: Module `jobqueue.src.jobqueue.workflow` not found in search paths
    at crates/ty_module_resolver/src/resolve.rs:215 on ThreadId(2)
    in ty_project::check_file with file=file(Id(c00))
    in ty_project::Project::check

  2026-08-24T04:12:00.383978Z DEBUG ty_module_resolver::resolve: Module `jobqueue.src.jobqueue.workflow` not found while looking in parent dirs (neither stub nor real module file)
    at crates/ty_module_resolver/src/resolve.rs:254 on ThreadId(2)
    in ty_project::check_file with file=file(Id(c00))
    in ty_project::Project::check

  2026-08-24T04:12:00.384037Z DEBUG ty_python_semantic::types::infer::builder: Relative module resolution `.` failed: could not resolve file `/tmp/ty-relative-import-repro.qy3Ut3/jobqueue/src/jobqueue/workflow.py` to a module (try adjusting configured search paths?)
    at crates/ty_python_semantic/src/types/infer/builder.rs:8488 on ThreadId(2)
    in ty_project::check_file with file=file(Id(c00))
    in ty_project::Project::check

  2026-08-24T04:12:00.418301Z DEBUG ty_module_resolver::resolve: Module `__builtins__` not found in search paths
    at crates/ty_module_resolver/src/resolve.rs:215 on ThreadId(2)
    in ty_project::check_file with file=file(Id(c00))
    in ty_project::Project::check

  2026-08-24T04:12:00.438647Z DEBUG ty_project: Checking all files took 0.056s
    at crates/ty_project/src/lib.rs:336 on ThreadId(2)
    in ty_project::Project::check

error[unresolved-import]: Cannot resolve imported module `.`
 --> jobqueue/src/jobqueue/workflow.py:1:1
  |
1 | from . import utils
  | ^^^^^^^^^^^^^^^^^^^
  |
info: Searched in the following paths during module resolution:
info:   1. /tmp/ty-relative-import-repro.qy3Ut3 (first-party code)
info:   2. vendored://stdlib (stdlib typeshed stubs vendored by ty)
info:   3. /tmp/ty-relative-import-repro.qy3Ut3/.pixi/envs/default/lib/python3.12/site-packages (site-packages)
info:   4. /tmp/ty-relative-import-repro.qy3Ut3/jobqueue/src (editable install)
info: make sure your Python environment is properly configured: https://docs.astral.sh/ty/modules/#python-environment
info: rule `unresolved-import` is enabled by default

Found 1 diagnostic
  2026-08-24T04:12:00.438961Z DEBUG ty: Exiting main loop
    at crates/ty/src/lib.rs:297 on ThreadId(1)
```

and here with the latest ty version (0.0.74)

```
  2026-08-24T04:20:36.884992Z DEBUG ty_module_resolver::resolve: Adding editable installation to module resolution path /tmp/ty-relative-import-repro.ZPUFln/jobqueue/src
    at crates/ty_module_resolver/src/resolve.rs:1019 on ThreadId(2)
    in ty_project::check_file with file=File(System("/tmp/ty-relative-import-repro.ZPUFln/jobqueue/src/jobqueue/workflow.py"))
    in ty_project::Project::check

  2026-08-24T04:20:36.885203Z DEBUG ty_module_resolver::resolve: Module `jobqueue.src.jobqueue.workflow` not found in search paths
    at crates/ty_module_resolver/src/resolve.rs:259 on ThreadId(2)
    in ty_project::check_file with file=File(System("/tmp/ty-relative-import-repro.ZPUFln/jobqueue/src/jobqueue/workflow.py"))
    in ty_project::Project::check

  2026-08-24T04:20:36.885316Z DEBUG ty_module_resolver::resolve: Module `jobqueue.src.jobqueue.workflow` not found while looking in parent dirs (typing mode)
    at crates/ty_module_resolver/src/resolve.rs:302 on ThreadId(2)
    in ty_project::check_file with file=File(System("/tmp/ty-relative-import-repro.ZPUFln/jobqueue/src/jobqueue/workflow.py"))
    in ty_project::Project::check

  2026-08-24T04:20:36.885388Z DEBUG ty_python_semantic::types::infer::builder::imports: Relative module resolution `.` failed: could not resolve file `/tmp/ty-relative-import-repro.ZPUFln/jobqueue/src/jobqueue/workflow.py` to a module (try adjusting configured search paths?)
    at crates/ty_python_semantic/src/types/infer/builder/imports.rs:303 on ThreadId(2)
    in ty_project::check_file with file=File(System("/tmp/ty-relative-import-repro.ZPUFln/jobqueue/src/jobqueue/workflow.py"))
    in ty_project::Project::check

  2026-08-24T04:20:36.915787Z DEBUG ty_module_resolver::resolve: Module `__builtins__` not found in search paths
    at crates/ty_module_resolver/src/resolve.rs:259 on ThreadId(2)
    in ty_project::check_file with file=File(System("/tmp/ty-relative-import-repro.ZPUFln/jobqueue/src/jobqueue/workflow.py"))
    in ty_project::Project::check

  2026-08-24T04:20:36.932994Z DEBUG ty_project: Checking all files took 0.049s
    at crates/ty_project/src/lib.rs:420 on ThreadId(2)
    in ty_project::Project::check

error[unresolved-import]: Cannot resolve imported module `.`
 --> jobqueue/src/jobqueue/workflow.py:1:1
  |
1 | from . import utils
  | ^^^^^^^^^^^^^^^^^^^
info: Searched in the following paths during module resolution:
info:   1. /tmp/ty-relative-import-repro.ZPUFln (first-party code)
info:   2. vendored://stdlib (stdlib typeshed stubs vendored by ty)
info:   3. /tmp/ty-relative-import-repro.ZPUFln/.pixi/envs/default/lib/python3.12/site-packages (site-packages)
info:   4. /tmp/ty-relative-import-repro.ZPUFln/jobqueue/src (editable install)
info: make sure your Python environment is properly configured: https://docs.astral.sh/ty/modules/#python-environment
info: rule `unresolved-import` is enabled by default

Found 1 diagnostic
  2026-08-24T04:20:36.933362Z DEBUG ty: Exiting main loop
    at crates/ty/src/lib.rs:338 on ThreadId(1)
```

If you need any additional input, let me know.

**Comment by MichaReiser** [comment]
The project structure is 

```
project/
├── pixi.toml
├── jobqueue/                         # No __init__.py
│   └── src/                          # Editable-install search root
│       ├── setup.py
│       ├── setup.cfg
│       └── jobqueue/                 # Actual Python package
│           ├── __init__.py
│           ├── utils.py
│           └── workflow.py           # from . import utils
└── .pixi/envs/default/lib/python3.12/site-packages/
    └── __editable__.dummy_jobqueue-0.1.0.pth
```


When we resolve `.` in `jobqueue/workflow`, there are two possible module names for `workflow.py`:

* `jobqueue.workflow`
* `jobqueue.src.jobqueue.workflow` (ty picks this one)


I think there's an argument that we should resolve the module name from the deepest matching search path instead of resolving them in order. Although, our default first party search path of `[".", "src"]` makes this a rather prominent change


mypy, pyrefly, and pyright resolve this layout correctly

**Comment by MichaReiser** [comment]
Hmm, this is a bit trickier and sort of reveals an issue that we resolve relative imports by going from `importing file` -> `absolute module name` -> `absolute module name for relative import` -> `resolve module`. 


We can fix the above issue by always using the deepest nested path, see my linked PR. However, there are other inconsistencies where ty resolves the wrong relative import:

`src/package/utils.py`:

```py
origin = "outer"

```

`src/nested/package/utils.py`:

```py
origin = "inner"

```

`src/nested/package/module.py`:

```py
from . import utils

```

`src/main.py`:

```py
import sys
from typing import reveal_type

# Match ty.json in the playground's Python runtime.
sys.path[:0] = ["/playground/src", "/playground/src/nested"]

from package.module import utils as short_utils
from nested.package.module import utils as long_utils

reveal_type(short_utils.origin)
reveal_type(long_utils.origin)
print(short_utils.origin, long_utils.origin)

```

`ty.json`:

```json
{
  "environment": {
    "root": [
      "/src",
      "/src/nested"
    ]
  }
}
```

[Playground](https://play.ty.dev/4bcbb749-a2ad-4cde-8486-ed3e3fe85b3e)
